File: cpp/problem12.cpp
/*
https://projecteuler.net/problem=12

Highly divisible triangular number

What is the value of the first triangle number to have over five hundred divisors?
*/

#include <chrono>
#include <iostream>
using namespace std;

struct result {
    long long value;
    long long time_ns;
};

int count_divisors_sqrt(long long n) {
    int count = 0;
    for (long long d = 1; d * d <= n; ++d) {
        if (n % d == 0) {
            count += 2;
            if (d * d == n) {
                --count;
            }
        }
    }
    return count;
}

int count_divisors_factorized(long long n) {
    int total = 1;
    int exp = 0;
    while (n % 2 == 0) {
        n /= 2;
        ++exp;
    }
    if (exp > 0) {
        total *= (exp + 1);
    }

    for (long long p = 3; p * p <= n; p += 2) {
        exp = 0;
        while (n % p == 0) {
            n /= p;
            ++exp;
        }
        if (exp > 0) {
            total *= (exp + 1);
        }
    }

    if (n > 1) {
        total *= 2;
    }

    return total;
}

long long evaluate_baseline() {
    const int target = 500;
    long long triangle = 0;

    for (long long n = 1;; ++n) {
        triangle += n;
        if (count_divisors_sqrt(triangle) > target) {
            return triangle;
        }
    }
}

long long evaluate_optimized() {
    const int target = 500;

    for (long long n = 1;; ++n) {
        long long a = n;
        long long b = n + 1;

        if (a % 2 == 0) {
            a /= 2;
        } else {
            b /= 2;
        }

        int divisors = count_divisors_factorized(a) * count_divisors_factorized(b);
        if (divisors > target) {
            return n * (n + 1) / 2;
        }
    }
}

result euler() {
    const auto start = std::chrono::steady_clock::now();
    long long value = evaluate_optimized();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

result euler_baseline() {
    const auto start = std::chrono::steady_clock::now();
    long long value = evaluate_baseline();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

int main() {
    std::cout << "Project Euler Problem 12" << std::endl;

    result baseline = euler_baseline();
    result optimized = euler();

    std::cout << "Result (baseline): " << baseline.value << std::endl;
    std::cout << "Time (baseline): " << baseline.time_ns << " ns" << std::endl;
    std::cout << "Result (optimized): " << optimized.value << std::endl;
    std::cout << "Time (optimized): " << optimized.time_ns << " ns" << std::endl;

    if (optimized.time_ns > 0) {
        double speedup = static_cast<double>(baseline.time_ns) / static_cast<double>(optimized.time_ns);
        std::cout << "Speedup: " << speedup << "x" << std::endl;
    }

    return 0;
}

File: scala/problem12.scala
/*
https://projecteuler.net/problem=12

Highly divisible triangular number
*/

object problem12 {
  private def divisorCount(n: Long): Int = {
    var value = n
    var factor = 2L
    var count = 1

    while (factor * factor <= value) {
      var power = 0
      while (value % factor == 0) {
        power += 1
        value /= factor
      }
      if (power > 0) {
        count *= (power + 1)
      }
      factor += 1
    }

    if (value > 1) count * 2 else count
  }

  def baselineAnswer: Long = {
    var i = 1L
    var triangle = 1L
    while (divisorCount(triangle) <= 500) {
      i += 1
      triangle += i
    }
    triangle
  }

  def optimizedAnswer: Long = {
    def triangleDivisorCount(n: Long): Int = {
      var a = n
      var b = n + 1L

      if (a % 2L == 0L) a /= 2L else b /= 2L

      divisorCount(a) * divisorCount(b)
    }

    var n = 1L
    while (triangleDivisorCount(n) <= 500) {
      n += 1L
    }

    n * (n + 1L) / 2L
  }

  case class Result(value: Long, timeNs: Long)

  def answer: Long = {
    optimizedAnswer
  }

  def eulerBaseline(): Result = {
    val start = System.nanoTime()
    val value = baselineAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def euler(): Result = {
    val start = System.nanoTime()
    val value = optimizedAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def main(args: Array[String]): Unit = {
    println("Project Euler Problem 12")

    val baseline = eulerBaseline()
    val optimized = euler()

    println(s"Result (baseline): ${baseline.value}")
    println(s"Time (baseline): ${baseline.timeNs} ns")
    println(s"Result (optimized): ${optimized.value}")
    println(s"Time (optimized): ${optimized.timeNs} ns")

    if (optimized.timeNs > 0) {
      val speedup = baseline.timeNs.toDouble / optimized.timeNs.toDouble
      println(s"Speedup: ${speedup}x")
    }
  }
}
